aten/src/ATen/native/quantized/cpu/qconv.cpp

```diff
@@ -244,6 +265,155 @@ class QConv2dInt8 final : public c10::OperatorKernel {
   }
+#endif
+#ifdef USE_PYTORCH_QNNPACK
+  at::Tensor qnnpack_conv(
+      Tensor act,
+      Tensor packed_weight,
+      torch::List<int64_t> stride,
+      torch::List<int64_t> padding,
+      torch::List<int64_t> dilation,
+      int64_t groups,
+      double output_scale,
+      int64_t output_zero_point) {
+    TORCH_CHECK(
+        act.ndimension() == 4,
+        "quanitzed::conv2d (qnnpack): Expected activation tensor to be 4-dimensional");
+    TORCH_CHECK(stride.size() == 2, "qnnpack_conv2d(): 2D convolution only");
+    TORCH_CHECK(
+        padding.size() == 2,
+        "quanitzed::conv2d (qnnpack): Specify top/left padding only. \
+        bottom/right padding assumed to be equal to top/left");
+    TORCH_CHECK(dilation.size() == 2, "quanitzed::conv2d (qnnpack): 2D convolution only");
+
+    PackedConvWeightsQnnp& pack_ptr =
+        cpp_custom_type_hack::cast<PackedConvWeightsQnnp>(packed_weight);
+    auto packB = pack_ptr.w.get();
+    auto& kernel = pack_ptr.kernel;
+    auto kernel_zp = pack_ptr.w_zp;
+    auto kernel_scale = pack_ptr.w_scale;
+
+    const uint32_t kernel_h = kernel[0];
+    const uint32_t kernel_w = kernel[1];
+    const auto out_ch = packB->getOutputChannels();
+    // inputs are in NHWC format
+    Tensor input_contig = act.contiguous();
+    int N = input_contig.size(0);
+    int H = input_contig.size(1);
+    int W = input_contig.size(2);
+    int in_ch = input_contig.size(3);
+    int K = out_ch; // output channels
+
+    uint32_t stride_h = stride[0];
+    uint32_t stride_w = stride[1];
+    uint32_t pad_t = padding[0];
+    uint32_t pad_l = padding[1];
+    uint32_t dilation_h = dilation[0];
+    uint32_t dilation_w = dilation[1];
+
+    auto output_min = ReluFused
+        ? activationLimits(output_scale, output_zero_point, Activation::RELU)
+              .first
+        : std::numeric_limits<uint8_t>::min();
+    auto output_max = ReluFused
+        ? activationLimits(output_scale, output_zero_point, Activation::RELU)
+              .second
+        : std::numeric_limits<uint8_t>::max();
+    qnnpack::conv_param_t conv_p(
+        {kernel_w, kernel_h},
+        {stride_w, stride_h},
+        {dilation_w, dilation_h},
+        {pad_t, pad_l, pad_t, pad_l},
+        groups,
+        in_ch,
+        out_ch,
+        kernel_zp,
+        kernel_scale,
+        output_min,
+        output_max);
+
+    auto outShape =
+        convOutputShape(N, H, W, K, kernel, stride, padding, dilation);
+    TORCH_CHECK(
+        std::all_of(
+            outShape.begin(), outShape.end(), [](int64_t i) { return i > 0; }),
+        "quanitzed::conv2d (qnnpack): each dimension of output tensor should be greater "
+        "than 0")
+    TORCH_CHECK(
+        (outShape[3] == out_ch),
+        "quanitzed::conv2d (qnnpack): Number of filters must be equal to number of "
+        "output channels")
+
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
```

Comment:
we don't support per channel quant. Do you have check about it somewhere?

Reply:
Yes, added a check in the prepack function.